src/primitives/block.h

```diff
@@ -136,23 +142,37 @@ class CBlockHeader
         READWRITE(hashMerkleRoot);
         READWRITE(nTime);
         READWRITE(nBits);
-        READWRITE(nNonce);
-        // Firo - MTP
-        // On read: allocate and read. On write: write only if already allocated
-        if (IsMTP()) {
-            READWRITE(nVersionMTP);
-            READWRITE(mtpHashValue);
-            READWRITE(reserved[0]);
-            READWRITE(reserved[1]);
-            if (ser_action.ForRead()) {
-                mtpHashData = make_shared<CMTPHashData>();
-                READWRITE(*mtpHashData);
-            }
-            else {
-                if (mtpHashData && !(s.GetType() & SER_GETHASH))
+
+        // Firo - ProgPoW
+        // Return std 4byte, if ProgPoW return 8byte
+        if (IsProgPow()) {
+
+            READWRITE(nHeight);
+            READWRITE(nNonce64);
```

Comment:
Normally in bitcoin when 32-bit `nOnce` is not enough (it never is) miner uses coinbase transaction. Because coinbase transaction input script is ignored you can put whatever you want there and it can serve as extension of `nOnce`. Why do you need to change it for PP and why 64-bit value is enough?

Reply:
I'm talking about something else

Once the nOnce range is exhausted (It could be possible even with 64 bits!) you'll want to add something to coinbase transaction modifying merkle tree hash and this in turn will change header hash and you'll be able to enumerate through all the nOnce possible values again.

You have to think about this possibility even if you move to 64 bits. Thus why change basic block structure if it doesn't completely solve your problem?

Not sure of typical hash rate with PP and how realistic it is to do more than 2^64 attempts. Bitcoin certainly does a LOT more than that.

Reply:
> (It could be possible even with 64 bits!)
On ethash (and ProgPoW) is impossible to exhaust the 64 bit range in a reasonable block time.
SHA256 and ethash are two completely different beasts: the latter is memory intensive (even more for PP) thus is way slower cause it's limited by memory bandwidth.
To give you some figures: an Antminer S19 is capable of 95TH, the best Nvidia 3090 can do 114MH on ethash and roughly 2/3 of that on ProgPow.

Reply:
If PP is slow enough not to exhaust 64-bit nOnce it won't be a problem.

Reply:
> Not sure of typical hash rate with PP and how realistic it is to do more than 2^64 attempts. Bitcoin certainly does a LOT more than that.

Not true ... do the math with a 95TH/s machine. It'd take two days to exhaust the 64bit range
`(1 << 64) / (95*(10**12)) / 86400 == 2 days`

Reply:
not on a single machine of course